File: MyMiniProject/angular-dotnet-finance/backend/Controllers/UsersController.cs
using FinanceApi.Data;
using FinanceApi.Models;
using Microsoft.AspNetCore.Authorization;
using Microsoft.AspNetCore.Mvc;
using Microsoft.EntityFrameworkCore;

namespace FinanceApi.Controllers;

[ApiController]
[Route("api/[controller]")]
[Authorize(Roles = "admin")]
public class UsersController(AppDbContext db) : ControllerBase
{
    [HttpGet]
    public async Task<ActionResult<List<UserDto>>> GetAll()
    {
        var users = await db.Users.OrderBy(u => u.Id).ToListAsync();
        return Ok(users.Select(AuthController.ToDto).ToList());
    }

    [HttpGet("{id:int}")]
    public async Task<ActionResult<UserDto>> Get(int id)
    {
        var user = await db.Users.FindAsync(id);
        if (user is null) return NotFound();
        return Ok(AuthController.ToDto(user));
    }

    [HttpPost]
    public async Task<ActionResult<UserDto>> Create([FromBody] CreateUserRequest req)
    {
        if (string.IsNullOrWhiteSpace(req.Username) || string.IsNullOrWhiteSpace(req.Password))
            return BadRequest(new { message = "Username และ Password จำเป็น" });

        if (await db.Users.AnyAsync(u => u.Username == req.Username))
            return Conflict(new { message = "Username นี้มีอยู่แล้ว" });

        var role = req.Role is "admin" or "user" ? req.Role : "user";
        var user = new User
        {
            Username = req.Username.Trim(),
            PasswordHash = BCrypt.Net.BCrypt.HashPassword(req.Password),
            FullName = req.FullName?.Trim() ?? req.Username,
            Email = req.Email?.Trim() ?? "",
            Role = role,
            IsActive = true
        };

        db.Users.Add(user);
        await db.SaveChangesAsync();
        return CreatedAtAction(nameof(Get), new { id = user.Id }, AuthController.ToDto(user));
    }

    [HttpPut("{id:int}")]
    public async Task<ActionResult<UserDto>> Update(int id, [FromBody] UpdateUserRequest req)
    {
        var user = await db.Users.FindAsync(id);
        if (user is null) return NotFound();

        user.FullName = req.FullName?.Trim() ?? user.FullName;
        user.Email = req.Email?.Trim() ?? user.Email;
        user.Role = req.Role is "admin" or "user" ? req.Role : user.Role;
        user.IsActive = req.IsActive;

        if (!string.IsNullOrWhiteSpace(req.Password))
            user.PasswordHash = BCrypt.Net.BCrypt.HashPassword(req.Password);

        await db.SaveChangesAsync();
        return Ok(AuthController.ToDto(user));
    }

    [HttpDelete("{id:int}")]
    public async Task<IActionResult> Delete(int id)
    {
        var user = await db.Users.FindAsync(id);
        if (user is null) return NotFound();

        if (user.Username == "admin")
            return BadRequest(new { message = "ห้ามลบบัญชี admin หลัก" });

        db.Users.Remove(user);
        await db.SaveChangesAsync();
        return NoContent();
    }
}

File: MyMiniProject/laravel-finance/app/Http/Controllers/Api/UsersController.php
<?php

namespace App\Http\Controllers\Api;

use App\Http\Controllers\Controller;
use App\Models\User;
use Illuminate\Http\Request;

class UsersController extends Controller
{
    public function index()
    {
        $users = User::orderBy('id')->get()->map->toApiArray()->values();

        return response()->json($users);
    }

    public function show(int $id)
    {
        $user = User::find($id);
        if (! $user) {
            return response()->json(['message' => 'Not found'], 404);
        }

        return response()->json($user->toApiArray());
    }

    public function store(Request $request)
    {
        $username = trim((string) $request->input('username', ''));
        $password = (string) $request->input('password', '');

        if ($username === '' || $password === '') {
            return response()->json(['message' => 'Username และ Password จำเป็น'], 400);
        }

        if (User::where('username', $username)->exists()) {
            return response()->json(['message' => 'Username นี้มีอยู่แล้ว'], 409);
        }

        $role = $request->input('role');
        $role = in_array($role, ['admin', 'user'], true) ? $role : 'user';

        $user = User::create([
            'username' => $username,
            'password' => $password,
            'full_name' => trim((string) ($request->input('fullName') ?: $username)),
            'email' => trim((string) $request->input('email', '')),
            'role' => $role,
            'is_active' => true,
        ]);

        return response()->json($user->toApiArray(), 201);
    }

    public function update(Request $request, int $id)
    {
        $user = User::find($id);
        if (! $user) {
            return response()->json(['message' => 'Not found'], 404);
        }

        if ($request->filled('fullName')) {
            $user->full_name = trim((string) $request->input('fullName'));
        }
        if ($request->has('email')) {
            $user->email = trim((string) $request->input('email', ''));
        }
        $role = $request->input('role');
        if (in_array($role, ['admin', 'user'], true)) {
            $user->role = $role;
        }
        if ($request->has('isActive')) {
            $user->is_active = filter_var($request->input('isActive'), FILTER_VALIDATE_BOOLEAN);
        }
        if ($request->filled('password')) {
            $user->password = $request->input('password');
        }

        $user->save();

        return response()->json($user->toApiArray());
    }

    public function destroy(int $id)
    {
        $user = User::find($id);
        if (! $user) {
            return response()->json(['message' => 'Not found'], 404);
        }

        if ($user->username === 'admin') {
            return response()->json(['message' => 'ห้ามลบบัญชี admin หลัก'], 400);
        }

        $user->delete();

        return response()->noContent();
    }
}
